k8srm-prototype/pkg/api/capacity_types.go

```diff
@@ -40,9 +40,17 @@ type DevicePoolSpec struct {
 	// +optional
 	Attributes []Attribute `json:"attributes,omitempty"`
 
-	// DeviceCount contains the total number of devices in the pool.
+	// Resources are pooled resources that are shared by all devices in the
+	// pool. This is typically used when representing a partitionable
+	// device, and need not be populated otherwise.
+	//
+	// +optional
+	Resources []ResourceCapacity `json:"resources,omitempty"`
+
+	// Devices contains the individual devices in the pool.
+	//
 	// +required
-	DeviceCount int `json:"count,omitempty"`
+	Devices []Device `json:"devices,omitempty"`
```

Comment:
Under this prototype, it seems that each GPU would be modeled as its own DevicePool.  So on a DGXA100 server, we would have 8 DevicePools, from which one could allocate either a full GPU or its mig devices. Is my understanding correct?

Reply:
Yeah. I think that's a better trade off than an extra level built into the pool. But we'll see what others'
 opinions are.

Reply:
Oh - you only need to do this is you have MIG or similar need for shared resources. So for you it would be pool per GPU, for other types of devices it would just be one pool.

Reply:
One other note. Adding different kinds of "Device" is equivalent to "a new structured model". So we can have an extension point there (e.g., "either Devices or WackyDevices must be populated") just like we do today.

Reply:
This really feels like the wrong place to be inserting this extension point. I'd actually expect it to be at a layer **above** `DevicePool`. (moving your `DriverName` and `NodeName` fields up there as well).

Reply:
I don't see a lot of value in another level, and keeping it here would mean the new model automatically gets the static attributes.

Reply:
But what if the new model doesn't want static attributes (or even worse, maybe it needs them defined slightly differently than we have defined them for this model)? It feels cleaner to me to have the extension point be at the highest level possible without making any assumption about what future models may want. If it does in fact want to use the exact same Attributes type (for example), then it can easily just include a field for that in its model (the type is already defined).

Reply:
I don't like the recursive model. I had proposed something similar during the KEP review for structured parameters, and it got (rightfully) shut down. The flat model is much better and more flexible.

One way to deal with what you describe is to decompose this into two separate models..... One model represents a single Device (analogous to our NamedResources model today) and we add a second model to represent DevicePools.

To me, this makes it much easier to convey to the driver developer what they need to do. If they have simple devices -- implement the standard Device model (following straightforward documentation for what's available in it). If they have a pool of devices, implement the DevicePool model (following straightforward documentation for what's available in this one instead).

Trying to pack every possible feature into a single model makes it harder for the driver developer to navigate in cases where they don't need all of the fancy features. For the claim side of things we need everything packed into one -- we don't want the end user having to use different constructs to request resources. But for the device vendor I think it's OK (even preferred) to clearly define separate models for them to implement that match the types of devices they want to represent.

Reply:
In the proposed model, a `DevicePool` is always a list of similar (not necessarily identical) devices, managed by the same driver. In a case in which there is no use of the `DevicePool.SharedResources` (the non-MIG case), a single DevicePool can hold all of the devices for a given driver for a given node, by using both `DevicePool.Attributes` and `Device.Attributes`. Or, the driver can split devices into pools by model (or whatever) and reduce or eliminate the use of `Device.Attributes`; it's up to the driver author and the claim and related allocation code won't care (it will effectively *ignore* the pools in the case of no `DevicePool.SharedResources`).

But once you need to use `SharedResources`, the driver will need to break up the devices in a node based upon how those shared resources are managed. In the MIG case, that means one DevicePool per card, with the partitions being the devices. In an SR-IOV NIC, you would probably need to do the same thing. Hmm. In fact, VFs live on PFs, much like MIGs live on memory slices, except that it's a 1-many relationships. What's not clear to me is if you can "consume" PFs directly or only VFs. In any case, suppose we have a card that has two PFs, each with two VFs (to keep it small...). I think again we would model (using the proposed model but adopting some names from the alternate in #8) that as a single physical card per pool, with the PFs and VFs looking something like partitions (assuming PFs are consumable):

```yaml
apiVersion: devmgmtproto.k8s.io/v1alpha1
kind: DevicePool
metadata:
  name: node-001-nic-001
spec:
  sharedResources:
    vfs-pf-0: 2
    vfs-pf-1: 2
  devices:
  - name: pf-0
    sharedResourceRequests:
      vfs-pf-0: 2
    resources:
      pf: 1
  - name: pf-1
    sharedResourceRequests:
      vfs-pf-1: 2
    resources:
      pf: 1
  - name: pf-0-vf-0
    sharedResourceRequests:
      vfs-pf-0: 1
    resources:
      vf: 1
  - name: pf-0-vf-1
    sharedResourceRequests:
      vfs-pf-0: 1
    resources:
      vf: 1
  - name: pf-1-vf-0
    sharedResourceRequests:
      vfs-pf-1: 1
    resources:
      vf: 1
  - name: pf-1-vf-1
    sharedResourceRequests:
      vfs-pf-1: 1
    resources:
      vf: 1
```

So, in this case I include per-device `resources`, because the assumption is that the user will claim either "pf"s or "vf"s. I don't know if that's a valid assumption. But we need to differentiate between the two. The other way to model that would be to have an `Attribute` that says if it is a VF or PF, and then the user would just request a "device", with the right constraint. They are equivalent...but which should we encourage?

Some more thoughts:
- Are most cases going to use some form of `SharedResources` or do we think there are lots of cases where it's just "whole PCIe" card?
- Are we trying to come up with a new abstraction that is already part of the PCIe structure of functions? Can we model based on that instead?
- If we have to do things like "pf-0-vfs" and "memory-slice-0", are we missing some fundamental part of the model around shared resources - like groups of shared resources a la Kevin's named model extension (I thought that was too complex...)?
- Maybe instead of `sharedResourceRequests` and `resources` we use `consumesSharedResources` and `providesClaimResources` to be super explicity on the differences between these things?

Reply:
Also - might vote is a pretty hard NO on the recursive model. Too complex.

Reply:
I think if a vendor _**only**_ has (1) simple devices OR (2) partitionable devices, then modeling devices using a `DevicePool` is a little ambiguous but it's not terrible.

Simple devices get smashed into one `DevicePool` object:
```
DevicePool:
  spec:
    driver: card.example.com
    devices:
    - name: card-0
      attributes: [ ... ]
    - name: card-1
      attributes: [ ... ]
    ...
```

Partitionable devices get split into separate `DevicePool` objects:
```
DevicePool:
  spec:
    driver: card.example.com
    sharedResources: [ ... ]
    devices:
    - name: card-0
      sharedResourceConsumption: [ ... ]
      attributes: [ ... ]
    - name: card-0-subcard0
      sharedResourceConsumption: [ ... ]
      attributes: [ ... ]
    - name: card-0-subcard1
      sharedResourceConsumption: [ ... ]
      attributes: [ ... ]
    ...
```
```
DevicePool:
  spec:
    driver: example.com/frobnicator
    sharedResources: [ ... ]
    devices:
    - name: card-1
      sharedResourceConsumption: [ ... ]
      attributes: [ ... ]
    - name: card-1-subcard0
      sharedResourceConsumption: [ ... ]
      attributes: [ ... ]
    - name: card-1
    - -subcard1
      sharedResourceConsumption: [ ... ]
      attributes: [ ... ]
    ...
```

Where it gets weird though is when a vendor has a mix of card types to support.
* Where does one draw the boundaries around each `DevicePool`?
* Do they put all of their simple devices in a single `DevicePool` alongside all partitionable devices in their own individual `DevicePool`s?
* Why is there a "pool" boundary at all if there is no real meaning tied to it?
* In the case of simple devices, is it only there so as to avoid having lots of separate `Device` objects in the API server?
* If we do want to support both simple and partitionable devices in a single `DevicePool`, do we need to add one level of embedding as @thockin proposed? An alternative of this is to go back to my concept of having a name for each `SharedResourceGroup`, forcing devices to refer back to the name of the group they pull a given shared resource from.  
* What happens as the size of these device pools grow and we hit the limit of a single API server object? Where do we draw the boundary then?

Reply:
I am going to open a new, unresolved comment :)

Reply:
https://github.com/kubernetes-sigs/wg-device-management/pull/5#discussion_r1591620848